### PR TITLE
Add geo_line aggregation

### LINKED ISSUE #41649: GeoLine Aggregation
### Feature

A metric aggregation that aggregates a set of points as 
a GeoJSON LineString ordered by some sort parameter.

#### specifics

A `geo_line` aggregation request would specify a `geo_point` field, as well
as a `sort` field. `geo_point` represents the values used in the LineString, 
while the `sort` values will be used as the total ordering of the points.

the `sort` field would support any numeric field, including date.

#### sample usage

```
{
	"query": {
		"bool": {
			"must": [
				{ "term": { "person": "004" } },
				{ "term": { "trajectory": "20090131002206.plt" } }
			]
		}
	},
	"aggs": {
		"make_line": {
			"geo_line": {
				"geo_point": {"field": "location"},
				"sort": { "field": "timestamp" }
			}
		}
	}
}
```

#### sample response

```
{
    "took": 21,
    "timed_out": false,
    "_shards": {...},
    "hits": {...},
    "aggregations": {
        "make_line": {
            "type": "LineString",
            "coordinates": [
                [
                    121.52926194481552,
                    38.92878997139633
                ],
                [
                    121.52922699227929,
                    38.92876998055726
                ],
             ]
        }
    }
}
```

#### visual response

Screen Shot 2019-04-26 at 9 40 07 AM

#### limitations

Due to the cardinality of points, an initial max of 10k points 
will be used. This should support many use-cases.

One solution to overcome this limitation is to keep a PriorityQueue of
points, and simplifying the line once it hits this max. If simplifying
makes sense, it may be a nice option, in general. The ability to use a parameter
to specify how aggressive one wants to simplify. This parameter could be 
the number of points. Example algorithm one could use with a PriorityQueue:
https://bost.ocks.org/mike/simplify/. This would still require O(m) space, where m
is the number of points returned. And would also require heapifying triangles
sorted by their areas, which would be O(log(m)) operations. Since sorting is done, 
anyways, simplifying would still be a O(n log(m)) operation, where n is the total number 
of points to filter........... something to explore

### PR DESCRIPTION
**Pull request by talevy**
A metric aggregation that aggregates a set of points as 
a GeoJSON LineString ordered by some sort parameter.

#### specifics

A `geo_line` aggregation request would specify a `geo_point` field, as well
as a `sort` field. `geo_point` represents the values used in the LineString, 
while the `sort` values will be used as the total ordering of the points.

the `sort` field would support any numeric field, including date.

#### sample usage

```
{
	"query": {
		"bool": {
			"must": [
				{ "term": { "person": "004" } },
				{ "term": { "trajectory": "20090131002206.plt" } }
			]
		}
	},
	"aggs": {
		"make_line": {
			"geo_line": {
				"point": {"field": "location"},
				"sort": { "field": "timestamp" },
                                "include_sort": true,
                                "sort_order": "desc",
                                "size": 15
			}
		}
	}
}
```

#### sample response

```
{
    "took": 21,
    "timed_out": false,
    "_shards": {...},
    "hits": {...},
    "aggregations": {
        "make_line": {
            "type": "LineString",
            "coordinates": [
                [
                    121.52926194481552,
                    38.92878997139633
                ],
                [
                    121.52922699227929,
                    38.92876998055726
                ],
             ]
        }
    }
}
```

#### visual response

Screen Shot 2019-04-26 at 9 40 07 AM

#### limitations

Due to the cardinality of points, an initial max of 10k points 
will be used. This should support many use-cases.

One solution to overcome this limitation is to keep a PriorityQueue of
points, and simplifying the line once it hits this max. If simplifying
makes sense, it may be a nice option, in general. The ability to use a parameter
to specify how aggressive one wants to simplify. This parameter could be 
the number of points. Example algorithm one could use with a PriorityQueue:
https://bost.ocks.org/mike/simplify/. This would still require O(m) space, where m
is the number of points returned. And would also require heapifying triangles
sorted by their areas, which would be O(log(m)) operations. Since sorting is done, 
anyways, simplifying would still be a O(n log(m)) operation, where n is the total number 
of points to filter........... something to explore

closes #41649

### CODE REVIEW

File: x-pack/plugin/geo/src/main/java/org/elasticsearch/xpack/geo/GeoLineAggregator.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        if (valuesSources == null) {
+            return buildEmptyAggregation();
+        }
+        List<InternalGeoLine.Position> bucketLine = positions.get(bucket);
+        if (bucketLine == null) {
+            bucketLine = Collections.emptyList();
+        } else {
+            Collections.sort(bucketLine);
+        }
+        return new InternalGeoLine(name, bucketLine, pipelineAggregators(), metaData());
```
Comment (polyfractal) [inline x-pack/plugin/geo/src/main/java/org/elasticsearch/xpack/geo/GeoLineAggregator.java]: This is probably over-optimizing but thought I'd mention:

Since the points are sorted by time, the x/y deltas should be reasonably small between consecutive points (in the general case).  We could, theoretically, re-compress the line using delta-encoding and it might save a bunch of space.  It would probably complicate the reduction on coordinating node though.

File: x-pack/plugin/geo/src/main/java/org/elasticsearch/xpack/geo/GeoLineAggregator.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+                    assert docGeoPointValues.docValueCount() == 1;
+                    final double sort = docSortValues.nextValue();
+                    final GeoPoint point = docGeoPointValues.nextValue();
+                    ArrayList<InternalGeoLine.Position> bucketLine = positions.get(bucket);
+                    if (bucketLine == null) {
+                        bucketLine = new ArrayList<>(1000);
+                    } else if (bucketLine.size() >= 10000) {
+                        throw new AggregationExecutionException("Cannot aggregate a line with more than 10000 points");
+                    }
+                    bucketLine.add(new InternalGeoLine.Position(
```
Comment (polyfractal) [inline x-pack/plugin/geo/src/main/java/org/elasticsearch/xpack/geo/GeoLineAggregator.java]: Was noodling over this and it'd be nice to avoid instantiating all these `Position` objects.  E.g. a dozen buckets with 10k points generates a lot of objects.

If we don't need the sort value to be a `double`, I'm wondering if perhaps we could use `PackedLongValues`, `PackedInts`, etc and encode the three values consecutively in the data-structure.  That'd avoid the object allocation, and gain the packing ability for maybe better compression.  

I suspect these won't pack well though, since I bet the lat/lon points requirea large number of bits per value.  But we could do similar with just a plain `long[]` and that'd also avoid the object allocations

The downside is that we'd have to allocate the whole block of memory up-front.  But I'm guessing most lines will be more-filled than less-filled so that's probably an ok trade-off.

File: x-pack/plugin/geo/src/main/java/org/elasticsearch/xpack/geo/GeoLineAggregator.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+                    // a real weight or a `missing` weight
+                    assert docSortValues.docValueCount() == 1;
+                    assert docGeoPointValues.docValueCount() == 1;
+                    final double sort = docSortValues.nextValue();
+                    final GeoPoint point = docGeoPointValues.nextValue();
+                    ArrayList<InternalGeoLine.Position> bucketLine = positions.get(bucket);
+                    if (bucketLine == null) {
+                        bucketLine = new ArrayList<>(1000);
+                    } else if (bucketLine.size() >= 10000) {
+                        throw new AggregationExecutionException("Cannot aggregate a line with more than 10000 points");
```
[the remark this answers is not in the document]
Comment (polyfractal) [reply x-pack/plugin/geo/src/main/java/org/elasticsearch/xpack/geo/GeoLineAggregator.java]: Regarding the 10k limit... seems like as good of a starting point as any.  That works out to `0.24mb` per shard per bucket theoreticall.  E.g. a 5-shard index will be something on the order of 1.2mb per bucket, which is on the heavier end for aggs but not unreasonable

File: x-pack/plugin/geo/src/main/java/org/elasticsearch/xpack/geo/InternalGeoLine.java [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        return builder;
+    }
+
+    @Override
+    public String toString() {
+        return Strings.toString(this);
+    }
+
+    @Override
+    public InternalAggregation doReduce(List<InternalAggregation> aggregations, ReduceContext reduceContext) {
```
Comment (polyfractal) [inline x-pack/plugin/geo/src/main/java/org/elasticsearch/xpack/geo/InternalGeoLine.java]: Potential optimization: since the shard results are all pre-sorted, we could divide-conquer pairwise merge the lists instead of dumping them all into a single list and sorting.  Or a priority queue implementation to track the lowest item from each list and sort them in a streaming fashion.

Might be overkill though, we're only talking about something like 10*10k points

File: x-pack/plugin/geo/src/main/java/org/elasticsearch/xpack/geo/GeoLineAggregator.java [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+import static org.elasticsearch.xpack.geo.GeoLineAggregationBuilder.SORT_FIELD;
+
+/**
+ * Metric Aggregation for computing the pearson product correlation coefficient between multiple fields
+ **/
+final class GeoLineAggregator extends MetricsAggregator {
+    /** Multiple ValuesSource with field names */
+    private final MultiValuesSource.AnyMultiValuesSource valuesSources;
+
+    private ObjectArray<ArrayList<InternalGeoLine.Position>> positions;
```
Comment (polyfractal) [inline x-pack/plugin/geo/src/main/java/org/elasticsearch/xpack/geo/GeoLineAggregator.java]: Followup from our slack chat about this... if we keep the arraylist implementation I think we'll have to manually account for bytes in the breaker instead of relying on the ObjectArray.  I don't think overriding `numBytesPerElement()` will work since each element (arraylist) will have a different size that needs tallying.  Instead each time an element is added to a list we'll need to bump the CB manually.  I think.

**Review by polyfractal — COMMENTED** [review]
Left a few comments about potential optimizations and such :)

Also a thought... I wonder if we could/should implement this as a pipeline agg?  It shares many similarities to metrics ordered by time.  A user could define a date_histogram on the time field, some kind of aggregating "metric" to collapse multiple geo_points in one bucket to a single point (average them?) and then the geo_line pipeline agg strings the multiple buckets into a single linestring.

It gives you all the sorting stuff for free, and sorta gives you line simplification out of the box, in that larger date_histo interval automatically gives you a less-granular line string.  Perhaps not "smart" in that it's simplifying by time and not line complexity, but might be an ok start?

Dunno, just a thought I had while looking over how it works.

**Comment by thomasneirynck** [comment]
Thanks, this is super useful!

The return-type is now a Geojson-Feature. Is this an issue from an aesthetic standpoint for Elasticsearch-API? (I don't know of any parallels where ES is hacking another data-format as an agg-response).

So just for some context: from a user-perspective (and specifically Maps), the idea to return valid GeoJson is mainly useful because it allows Maps to reuse the coordinates-array by-reference, without any post-processing (except for parsing of the JSON-response, which is a native browser function).

If it feels "odd" to wrap the `sort_values` inside the `Feature#properties`, I think it would be fine to just return the geometry-person of the ES-response as either:
- a valid GeoJson linestring
- a valid 2D GeoJson coordinate-array (array of array coordinate-pairs).

Basically, as long as the coordinates-array is there in valid GeoJson, it's a win for Maps.

The additional metadata about the individual coordinates (ie. `sort_values`) can always be in a separate object.

Just a thought, because if most users would exclude the sort_values from their agg, it might feel odd to just have dangling empty properties there.

----

Would it be useful to make it explicit if the line-string is complete or not in the response? Right now, we can compare doc-count of the bucket with point-count in the line, but this will no longer work when simplification would be introduced.

**Comment by talevy** [comment]
heya @thomasneirynck. thanks for the comments/concerns/suggestions!

> If it feels "odd" to wrap the sort_values inside the Feature#properties, I think it would be fine to just return the geometry-portion of the ES-response as either:

it does not feel odd to me, as this is a property of the geometry, just happens to be a multi-dimensional property :)

there is an `include_sort` flag that will hide or show the sort values, since sometimes the geometry itself is all that matters.

> Just a thought, because if most users would exclude the sort_values from their agg, it might feel odd to just have dangling empty properties there.

the properties will not be dangling. There is one other property that exists. the `complete` property that is true if the line returned is a complete representation of the data or if it had to drop some points.

> Would it be useful to make it explicit if the line-string is complete or not in the response?

yes, the `complete` exists for this purpose. I intend for there to be a `simplify` flag for the request to tell the aggregator whether to simplify the line returned or not. A line can be both simplified and complete, in the sense that all of the data was taken as input to the simplification algorithm

File: server/src/main/java/org/elasticsearch/search/aggregations/support/MultiValuesSource.java [hunk 6]
```diff
@@ -55,6 +57,46 @@ public SortedNumericDoubleValues getField(String fieldName, LeafReaderContext ct
         }
     }
 
+    public static class AnyMultiValuesSource extends MultiValuesSource<ValuesSource> {
+        public AnyMultiValuesSource(Map<String, ValuesSourceConfig> valuesSourceConfigs, QueryShardContext context) {
```
Comment (iverase) [inline server/src/main/java/org/elasticsearch/search/aggregations/support/MultiValuesSource.java:61]: maybe we should just call it `GeoLineMultiValueSource`? It is only used by that aggregation which expects a numeric field and a geo point field.

File: x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/InternalGeoLine.java [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+        out.writeBoolean(includeSorts);
+        sortOrder.writeTo(out);
+    }
+
+    @Override
+    public InternalAggregation reduce(List<InternalAggregation> aggregations, ReduceContext reduceContext) {
+        int mergedSize = 0;
+        boolean complete = true;
+        boolean includeSorts = true;
+        for (InternalAggregation aggregation : aggregations) {
```
Comment (iverase) [inline x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/InternalGeoLine.java:71]: As I understand, points on each aggregation are already sorted, I wonder if we should use a priority queue to retrieve the top points needed instead of copying all points into a new array?

**Review by iverase — COMMENTED** [review]
I have a first look. I like that when we build internal aggregations from the aggregator we sort the data. I wonder if we should use that in the reduce phase so instead of copying all points into a new array, we can use a priority queue to construct the new Internal Aggregation?

File: server/src/main/java/org/elasticsearch/search/aggregations/support/MultiValuesSource.java (hunk 6 above)
Comment (talevy) [reply server/src/main/java/org/elasticsearch/search/aggregations/support/MultiValuesSource.java:61]: great idea. will change it!

File: x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/InternalGeoLine.java (hunk 7 above)
Comment (talevy) [reply x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/InternalGeoLine.java:71]: I initially thought this type of optimization was unnecessary, but now that you bring it up, If we are aggregating against a lot of shards, then the  reduce can blow up to be quiet a lot of additional memory here. Will update to do this!

Comment (talevy) [reply x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/InternalGeoLine.java:76]: in my latest commit, I've switched to this strategy. I'm still debugging bugs in it. hope to resolve it soon!

Comment (talevy) [reply x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/InternalGeoLine.java:76]: oh my. strategy worked. it was my test that had a bug. resolved!

**Comment by talevy** [comment]
run elasticsearch-ci/2

File: x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/MergedGeoLines.java [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+        return finalSortValues;
+    }
+
+    /**
+     * merges <code>geoLines</code> into one sorted list of values representing the combined line.
+     */
+    public void merge() {
+        // 1. add first element of each sub line to heap
+        for (int i = 0; i < geoLines.size(); i++) {
+            add(i, 0);
```
Comment (iverase) [inline x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/MergedGeoLines.java:53]: What happen here if one of the geoLines has size == 0?

**Review by iverase — APPROVED** [review]
lgtm Just left a comment but not sure if it is possible

File: x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/MergedGeoLines.java (hunk 8 above)
Comment (talevy) [reply x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/MergedGeoLines.java:53]: good point. I will verify this. I was operating under the assumption that a new `InternalAggregation` is only made if the bucket is not empty, but that may not be the case for all parent bucketing aggregations
